classes/local/rule_instances_controller.php

```diff
@@ -168,18 +160,61 @@ public function get_rule_instance_records(): array {
     }
 
     /**
-     * Return the rule instance record matching the given instanceid.
+     * Return a list of hydrated rule instance objects.
+     *
+     * @return rule\rule_interface[]|rule\plugin_configurable[]|rule\instance_configurable[]
+     */
+    public function get_rule_instances(): array {
+        $instances = [];
+        foreach ($this->get_rule_instance_records() as $record) {
+            // Create a new rule instance object and add it to the list of instances.
+            $ruleinstance = rule\rule_factory::create_instance_from_record($record);
+            $instances[$ruleinstance->get_id()] = $ruleinstance;
+        }
+
+        return $instances;
+    }
+
+    /**
+     * Return a rule instance object matching the given instanceid.
      *
      * @param int $instanceid
-     * @return \stdClass|null A single rule instance record.
+     * @return rule\rule_interface A single rule instance record.
      */
-    public function get_rule_instance_by_id(int $instanceid) {
-        $instance = array_column($this->ruleinstances, null, 'id')[$instanceid];
-        if (!$instance) {
+    public function get_rule_instance_by_id(int $instanceid): rule\rule_interface {
+        $instances = $this->get_rule_instances();
+        if (!isset($instances[$instanceid])) {
             throw new coding_exception('Invalid instance ID');
         }
 
-        return $instance;
+        return $instances[$instanceid];
+    }
+
+    /**
+     * Return a list of active rule instance objects, excluding those where either the
+     * rule plugin or instance are disabled, or the rule plugin has not been configured.
+     *
+     * @return array
```

Comment:
(OK seems I lied above, this one can also be more precise. :joy:)

```suggestion
     * @return rule\rule_interface[]
```

Reply:
Yes these two methods should really return the same thing, only difference is `get_active_rule_instances()` filters out "inactive" instances. 

I.e. those where either the rule plugin or instance are disabled, or the rule plugin has not been configured.

Reply:
Ahh I remember why I used a union type in the docblock for `get_rule_instances()`, frustratingly vscodium isn't able to infer the correct type after the `is_subclass_of()` check in `get_active_rule_instances()`.

Lemme have a think.

Reply:
I guess we may have to accept that vscode can't do this and put up with , at least I can't think of a way to express it in the docblock.